### PR TITLE
KAFKA-3266: Implement KIP-140 RPCs and APIs for creating, altering, and listing ACLs

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/common/KafkaFuture.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
      * A function which takes objects of type A and returns objects of type B.
      */
     public static abstract class Function<A, B> {
-        public abstract B apply(A a);
+        public abstract B apply(A a) throws Throwable;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/KafkaFuture.java]: Not sure about this. Note that Java 8's Function doesn't specify any exceptions.

File: clients/src/main/java/org/apache/kafka/common/errors/InvalidResponseException.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+/**
+ * The response was invalid.
+ */
+public class InvalidResponseException extends ApiException {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/errors/InvalidResponseException.java]: Are we throwing this back to the users? What are they expected to do in this case?

File: clients/src/main/java/org/apache/kafka/common/requests/ListAclsResponse.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        this.throttleTimeMs = throttleTimeMs;
+        this.throwable = throwable;
+        this.acls = acls;
+    }
+
+    public ListAclsResponse(Struct struct) {
+        this.throttleTimeMs = struct.getInt(THROTTLE_TIME_MS);
+        short errorCode = struct.getShort(ERROR_CODE);
+        String errorMessage = struct.getString(ERROR_MESSAGE);
+        if (errorCode != 0) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/requests/ListAclsResponse.java]: Nit: we typically do the `Errors.forCode` inline with the `getShort` and then check against `NONE` instead of `0`

File: clients/src/main/java/org/apache/kafka/common/errors/SecurityDisabledException.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+ */
+package org.apache.kafka.common.errors;
+
+/**
+ * An error indicating that security is disabled on the broker.
+ */
+public class SecurityDisabledException extends ApiException {
+    private static final long serialVersionUID = 1L;
+
+    public SecurityDisabledException() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/errors/SecurityDisabledException.java]: Maybe it's worth not including this constructor. It's only used in tests and it's generally a good idea to provide a message with the exception.

File: checkstyle/import-control.xml [hunk 5]
```diff
@@ -110,6 +110,7 @@
     <subpackage name="requests">
       <allow pkg="org.apache.kafka.common.protocol" />
       <allow pkg="org.apache.kafka.common.network" />
+      <allow pkg="org.apache.kafka.common.requests" />
```
Comment (ijuma) [inline checkstyle/import-control.xml]: Hmm, is this really needed?

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java [hunk 6]
```diff
@@ -183,4 +184,64 @@ public ApiVersionsResults apiVersions(Collection<Node> nodes) {
...
     public abstract ApiVersionsResults apiVersions(Collection<Node> nodes, ApiVersionsOptions options);
+
+    /**
+     * Similar to #{@link AdminClient#listAcls(AclFixture, ListAclsOptions),
+     * but uses the default options.
+     *
+     * @param filter            The filter to use.
+     * @return                  The DeleteAclsResult.
+     */
+    public ListAclsResults listAcls(AclFixture filter) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: Hmm, seems like we are exposing a class from an internal package (`requests`) in a public class.

File: clients/src/main/java/org/apache/kafka/common/requests/AclFixture.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.requests;
+
+import java.util.Objects;
+
+public class AclFixture {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/requests/AclFixture.java]: I've only seen `Fixture` used in the context of tests. Have you seen it used in the context of ACLs before?

File: core/src/main/scala/kafka/security/auth/PermissionType.scala [hunk 8]
```diff
@@ -17,20 +17,25 @@
...
+import org.apache.kafka.common.requests.AclPermissionType
 
 /**
  * PermissionType.
  */
 
 
-sealed trait PermissionType extends BaseEnum
+sealed trait PermissionType extends BaseEnum {
+  val toJava : AclPermissionType
```
Comment (ijuma) [inline core/src/main/scala/kafka/security/auth/PermissionType.scala]: Nit: there should be no space before the colon.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 9]
```diff
@@ -1463,6 +1470,223 @@ class KafkaApis(val requestChannel: RequestChannel,
     sendResponseMaybeThrottle(request, createResponse)
   }
 
+  def handleListAcls(request: RequestChannel.Request): Unit = {
+    authorizeClusterAction(request)
+    val listAclsRequest = request.body[ListAclsRequest]
+    authorizer match {
+      case None => {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: Nit: the brace between match clauses is never necessary.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 10]
```diff
@@ -1463,6 +1470,223 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    fixture.resource().resourceType() match {
+      case AclResourceType.ANY => return Failure(new InvalidRequestException("Invalid resource type"))
+      case AclResourceType.UNKNOWN => return Failure(new InvalidRequestException("Invalid resource type"))
+      case _ => ()
+    }
+    var resourceType: ResourceType = null
+    try {
+      resourceType = ResourceType.fromJava(fixture.resource().resourceType)
+    } catch {
+      case throwable: Throwable => return Failure(new InvalidRequestException("Unknown resource type"))
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: This is a bit of a big hammer. It seems more reasonable for `fromJava` to return an `Option`. Also, we should probably include the value in the exception message. Same for any other "Invalid..." exception.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR. I did an initial pass and left some comments. Will continue reviewing.

As an aside, this is almost +3k lines of code, more than I expected. :)

File: clients/src/main/java/org/apache/kafka/common/KafkaFuture.java (hunk 1 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/KafkaFuture.java]: That's a good point.  It might be good to avoid throwing checked exceptions from Function#apply, to make it easier to transition to Java 8 when it becomes available.

File: clients/src/main/java/org/apache/kafka/common/errors/InvalidResponseException.java (hunk 2 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/errors/InvalidResponseException.java]: Fair enough.  Let's just use `UnknownServerException`.

File: checkstyle/import-control.xml (hunk 5 above)
Comment (cmccabe) [inline checkstyle/import-control.xml]: Otherwise it whines about importing inner classes from the same package.

File: clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java (hunk 6 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AdminClient.java]: Good point.  I will move these classes into `org.apache.kafka.clients.admin`, since they are intended to be used by end-users.

File: clients/src/main/java/org/apache/kafka/common/requests/AclFixture.java (hunk 7 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/requests/AclFixture.java]: Fixture is just a generic word for an emplacement, nothing to do with tests per se.  In this case it seems appropriate since it represents an ACL that has been installed on some resource.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 9 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Hmm, can you be a bit clearer which brace is unnecessary?  I get syntax highlighting errors if I delete any of them.

File: checkstyle/import-control.xml (hunk 5 above)
Comment (cmccabe) [reply checkstyle/import-control.xml]: yes.  otherwise, it complains about attempts to import inner classes from the same package.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 9 above)
Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]:
```scala
foo match {
  case None => {
    ...
  }
  case Some(x) =>...
}
```
can be written as even if there are multiple expressions in :
```scala
foo match {
  case None =>
    ...
  case Some(x) =>...
```

Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Thanks.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 10 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: OK.  I can convert them to return Try[...].  The exception text will include the value and what the possible valid values are.

File: core/src/main/scala/kafka/security/auth/Operation.scala [hunk 11]
```diff
@@ -66,8 +68,12 @@ object Operation {
...
    }
 
-  def fromJava(operation: AclOperation): Operation = {
-    fromString(operation.toString)
+  def fromJava(operation: AclOperation): Try[Operation] = {
+    try {
+      Success(fromString(operation.toString))
+    } catch {
+      case throwable: Throwable => Failure(throwable)
+    }
```
Comment (ijuma) [inline core/src/main/scala/kafka/security/auth/Operation.scala:76]: the `Try` constructor does the `try/catch` for you. Same for other methods that do the same.

**Comment by cmccabe** [comment]
retest this please

File: checkstyle/suppressions.xml [hunk 12]
```diff
@@ -40,20 +40,23 @@
               files=".*/protocol/Errors.java"/>
 
     <suppress checks="BooleanExpressionComplexity"
-              files="(Utils|KafkaLZ4BlockOutputStream).java"/>
+              files="(Utils|KafkaLZ4BlockOutputStream|AclData).java"/>
```
Comment (ijuma) [inline checkstyle/suppressions.xml:43]: I think we don't have `AclData` any more.

File: checkstyle/import-control.xml [hunk 13]
```diff
@@ -108,8 +108,10 @@
     </subpackage>
 
     <subpackage name="requests">
+      <allow pkg="org.apache.kafka.clients.admin" />
```
Comment (ijuma) [inline checkstyle/import-control.xml:111]: We should check if we can remove this after KAFKA-5265.

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+public class AccessControlEntry {
+    final AccessControlEntryData data;
+
+    public AccessControlEntry(String principal, String host, AclOperation operation, AclPermissionType permissionType) {
+        Objects.requireNonNull(principal);
+        Objects.requireNonNull(host);
+        Objects.requireNonNull(operation);
+        assert operation != AclOperation.ANY;
+        Objects.requireNonNull(permissionType);
+        assert permissionType != AclPermissionType.ANY;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java:35]: Since asserts are not enabled in any environment at the moment, can we just do plain checks? Also, it's a bit nicer if we include a brief error message.

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+package org.apache.kafka.clients.admin;
+
+import java.util.Objects;
+
+/**
+ * Represents an access control entry.  ACEs are a tuple of principal, host,
+ * operation, and permissionType.
+ */
+public class AccessControlEntry {
+    final AccessControlEntryData data;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java:27]: Is it worth having this class? Not sure there is much overlap apart from storing the fields.

File: clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+
+package org.apache.kafka.clients.admin;
+
+import java.util.HashMap;
+import java.util.Locale;
+
+/**
+ * Represents an operation which an ACL grants or denies permission to perform.
+ */
+public enum AclOperation {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:26]: Another reason for having these classes in common (i.e. KAFKA-5265) is that they can potentially be used by the Authorizer interface when we move it to Java.

File: clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+        AclOperation operation = CODE_TO_VALUE.get(code);
+        if (operation == null) {
+            return UNKNOWN;
+        }
+        return operation;
+    }
+
+    private final byte code;
+
+    AclOperation(byte code) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:111]: Other enums in Kafka tend to use id for this.

File: clients/src/main/java/org/apache/kafka/clients/admin/CreateAclsResults.java [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+ */
+public class CreateAclsResults {
+    private final Map<AclBinding, KafkaFuture<Void>> futures;
+
+    CreateAclsResults(Map<AclBinding, KafkaFuture<Void>> futures) {
+        this.futures = futures;
+    }
+
+    /**
+     * Return a map from topic names to futures which can be used to check the status of
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/CreateAclsResults.java:35]: This comment needs to be updated.

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public String toString() {
+        return data.toString();
+    }
+
+    /**
+     * Return true if this AclResource has any UNKNOWN components.
+     */
+    public boolean unknown() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java:70]: Would `isUnknown` be clearer? I find that boolean methods without any prefix feel a bit ambiguous when reading them.

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryFilter.java [hunk 20]
```diff
@@ -0,0 +1,117 @@
...
+    public boolean matches(AccessControlEntry other) {
+        if ((principal() != null) && (!data.principal().equals(other.principal())))
+            return false;
+        if ((host() != null) && (!host().equals(other.host())))
+            return false;
+        if ((operation() != AclOperation.ANY) && (!operation().equals(other.operation())))
+            return false;
+        if ((permissionType() != AclPermissionType.ANY) && (!permissionType().equals(other.permissionType())))
+            return false;
+        return true;
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryFilter.java:86]: It seems like the logic would be simpler if we made them all positive, no?

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryData.java [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+    AclPermissionType permissionType() {
+        return permissionType;
+    }
+
+    /**
+     * Returns a string describing an ANY or UNKNOWN field, or null if there is
+     * no such field.
+     */
+    public String findIndefiniteField() {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryData.java:59]: Hmm, we should probably specify that we can change the output of this String (i.e. parsing may break). I would have preferred if we didn't expose this as a public method and had a utility for it.

File: clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * CLUSTER_ACTION operation.
+     */
+    CLUSTER_ACTION((byte) 9);
+
+    private final static HashMap<Byte, AclOperation> CODE_TO_VALUE = new HashMap<>();
+
+    static {
+        for (AclOperation operation : AclOperation.values()) {
+            CODE_TO_VALUE.put(operation.code, operation);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:82]: Since the code and enum id coincide exactly, is it worth having this map? Maybe we can add it later, if needed and simply look for it in the `values()` array.

File: clients/src/main/java/org/apache/kafka/clients/admin/DeleteAclsResults.java [hunk 23]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    private final Map<AclBindingFilter, KafkaFuture<FilterResults>> futures;
+
+    DeleteAclsResults(Map<AclBindingFilter, KafkaFuture<FilterResults>> futures) {
+        this.futures = futures;
+    }
+
+    /**
+     * Return a map from topic names to futures which can be used to check the status of
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/DeleteAclsResults.java:70]: Needs to be updated.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 24]
```diff
@@ -819,12 +833,12 @@ void call(Call call, long now) {
     public CreateTopicResults createTopics(final Collection<NewTopic> newTopics,
                                            final CreateTopicsOptions options) {
         final Map<String, KafkaFutureImpl<Void>> topicFutures = new HashMap<>(newTopics.size());
-        for (NewTopic newTopic : newTopics) {
-            topicFutures.put(newTopic.name(), new KafkaFutureImpl<Void>());
-        }
         final Map<String, CreateTopicsRequest.TopicDetails> topicsMap = new HashMap<>(newTopics.size());
         for (NewTopic newTopic : newTopics) {
-            topicsMap.put(newTopic.name(), newTopic.convertToTopicDetails());
+            if (topicFutures.get(newTopic.name()) == null) {
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:838]: Can't we just `put`? If there are duplicates (which is rare), we will simply overwrite, no? Same question for other places where we do the same.

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java [hunk 25]
```diff
@@ -957,16 +973,20 @@ void handleFailure(Throwable throwable) {
...
+            }
         }
         final long now = time.milliseconds();
         runnable.call(new Call("describeTopics", calcDeadlineMs(now, options.timeoutMs()),
             new ControllerNodeProvider()) {
 
             @Override
             AbstractRequest.Builder createRequest(int timeoutMs) {
-                return new MetadataRequest.Builder(new ArrayList<>(topicNames));
+                return new MetadataRequest.Builder(topicNamesList);
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:989]: Out of curiosity, why was this changed?

File: clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+class RequestUtils {
+    static Resource resourceFromStructFields(Struct struct) {
+        byte resourceType = struct.getByte("resource_type");
+        String name = struct.getString("resource_name");
+        return new Resource(ResourceType.fromCode(resourceType), name);
+    }
+
+    static void resourceSetStructFields(Resource resource, Struct struct) {
+        struct.set("resource_type", resource.resourceType().code());
+        struct.set("resource_name", resource.name());
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java:37]: It would be good to have constants instead of hardcoding the fields in many places.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 27]
```diff
@@ -1655,6 +1663,217 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+
+  /**
+    * Convert an ACL binding filter to a Scala object.
+    * All ACL and resource fields must be specified (no UNKNOWN, ANY, or null fields are allowed.)
+    *
+    * @param filter     The binding filter as a Java object.
+    * @return           The binding filter as a scala object, or an exception if there was an error
+    *                   converting the Java object.
+    */
+  def toScala(filter: AclBindingFilter) : Try[(Resource, Acl)] = {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1705]: Can we move some of this logic to a class in the `kafka.security.auth` package? Generally, we should do a better job of not cluttering `KafkaApis` too much since it needs to handle every protocol supported by the system.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 28]
```diff
@@ -1655,6 +1663,217 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+
+  /**
+    * Convert a Scala ACL binding to a Java object.
+    *
+    * @param acl        The binding as a Scala object.
+    * @return           The binding as a Java object.
+    */
+  def toJava(acl: (Resource, Acl)) : AclBinding = {
+    acl match {
+      case (resource, acl) =>
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1753]: If there's a single `case`, there's no reason to use `match`. You can simply do:

```scala
val (resource, acl) = acl
```

Although there's a name clash, so maybe the parameter should be `aclBinding` or something.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 29]
```diff
@@ -1655,6 +1663,217 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+    authorizer match {
+      case None =>
+        def createResponse(throttleTimeMs: Int): AbstractResponse =
+          createAclsRequest.getErrorResponse(throttleTimeMs,
+            new SecurityDisabledException("No Authorizer is configured on the broker."))
+        sendResponseMaybeThrottle(request, createResponse)
+      case Some(auth) =>
+        val errors = mutable.HashMap[Int, Throwable]()
+        var creations = ListBuffer[(Resource, Acl)]()
+        for (i <- 0 to createAclsRequest.aclCreations().size() - 1) {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1773]: You can use `until` instead of `to` to avoid the `-1` at the end.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 30]
```diff
@@ -1655,6 +1663,217 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+        for (i <- 0 to deleteAclsRequest.filters().size - 1) {
+          toDelete.put(i, new ListBuffer[(Resource, Acl)]())
+        }
+        if (deleteAclsRequest.filters().asScala.exists { f => !f.matchesAtMostOne() }) {
+          // Delete based on filters that may match more than one ACL.
+          val aclMap : Map[Resource, Set[Acl]] = auth.getAcls()
+          aclMap.foreach {
+            case (resource, acls) => {
+              acls.foreach {
+                case (acl) => {
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1824]: There's no need for `case` here. Also, the parameter can be in the previous line and we don't need the `{` after `=>`.

File: core/src/test/scala/integration/kafka/api/KafkaAdminClientIntegrationTest.scala [hunk 31]
```diff
@@ -156,6 +156,25 @@ class KafkaAdminClientIntegrationTest extends KafkaServerTestHarness with Loggin
...
+   */
+  @Test
+  def testAclOperations(): Unit = {
+    client = AdminClient.create(createConfig())
+    assertFutureExceptionTypeEquals(client.describeAcls(AclBindingFilter.ANY).all(), classOf[SecurityDisabledException])
+    assertFutureExceptionTypeEquals(client.createAcls(Collections.singleton(ACL1)).all(),
+        classOf[SecurityDisabledException])
+    assertFutureExceptionTypeEquals(client.deleteAcls(Collections.singleton(ACL1.toFilter())).all(),
+      classOf[SecurityDisabledException])
+    client.close()
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/KafkaAdminClientIntegrationTest.scala:175]: We don't need this since it's done in tearDown.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala [hunk 32]
```diff
@@ -35,5 +44,77 @@ class SaslSslAdminClientIntegrationTest extends KafkaAdminClientIntegrationTest
...
+
+  def waitForDescribeAcls(client: AdminClient, filter: AclBindingFilter, acls: Set[AclBinding]): Unit = {
+    TestUtils.waitUntilTrue(() => {
+      val results = client.describeAcls(filter).all().get()
+      acls == results.asScala.toSet
+    }, "timed out waiting for ACLs")
+  }
+
+  @Test
+  def testAclOperations2(): Unit = {
```
Comment (ijuma) [inline core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala:86]: It would be good to explain how this differs from the other, `2` doesn't give it away.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 33]
```diff
@@ -1655,6 +1663,217 @@ class KafkaApis(val requestChannel: RequestChannel,
...
+                    new AccessControlEntry(acl.principal.toString(), acl.host.toString(), acl.operation.toJava, acl.permissionType.toJava))
+                if (filter.matches(fixture))
+                  returnedAcls.add(fixture)
+              }
+            }
+          }
+        }
+        def createResponse(throttleTimeMs: Int): AbstractResponse =
+          new DescribeAclsResponse(throttleTimeMs, null, returnedAcls)
+        sendResponseMaybeThrottle(request, createResponse)
```
Comment (ijuma) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1693]: Isn't it easier to just inline `createResponse`? eg:

```scala
sendResponseMaybeThrottle(request, throttleTimeMs => new DescribeAclsResponse(throttleTimeMs, null, returnedAcls))
```
A few others are the same.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, looks good overall. I left a few minor comments. Some of them can be addressed subsequently as well.

File: clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java (hunk 16 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:26]: good point

File: clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java (hunk 22 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:82]: Calling `Enum#values` instantiates a new array each time, which is inefficient.

Comment (ijuma) [reply clients/src/main/java/org/apache/kafka/clients/admin/AclOperation.java:82]: I know, but you can save it in a static final.

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryData.java (hunk 21 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryData.java:59]: I'll add an Unstable annotation to the function.

Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntryData.java:59]: Also the ACEData function should be package-private.  Fixed

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 24 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:838]: We want to make sure duplicates don't get added to the list of topics to send

File: clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java (hunk 26 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/common/requests/RequestUtils.java:37]: This at least centralizes it in this class.  But yes, I agree, a constant would be good

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 33 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1693]: agree

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 27 above)
Comment (cmccabe) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1705]: yeah

**Review by ijuma — APPROVED** [review]
Synced with @cmccabe.

Since the suggestions don't affect correctness, the feature freeze is near and we also need to merge KIP-133 after this, I will go ahead and merge the branch as it is. Colin will file a JIRA and submit the improvements that he agrees with in a subsequent PR.

File: core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala (hunk 32 above)
Comment (cmccabe) [reply core/src/test/scala/integration/kafka/api/SaslSslAdminClientIntegrationTest.scala:86]: maybe testAclDeleteByFIlter

File: clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java (hunk 25 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:989]: it's part of the change above to handle duplicates

File: clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java (hunk 15 above)
Comment (cmccabe) [reply clients/src/main/java/org/apache/kafka/clients/admin/AccessControlEntry.java:27]: would prefer to have it, but don't feel strongly.  it's non-public
